### PR TITLE
Implementation of PowerTransform

### PR DESCRIPTION
**Pull request by vishwakftw**
This closes https://github.com/probtorch/pytorch/issues/114. 

cc: @fritzo @alicanb

### CODE REVIEW

File: torch/distributions/transforms.py [hunk 1]
```diff
@@ -298,9 +301,60 @@ def log_abs_det_jacobian(self, x, y):
...
+        else:
+            if other.exponent == self.exponent:
+                return True
+            else:
+                return False
+
+    def _call(self, x):
+        return x.pow(self.exponent)
+
+    def _inverse(self, y):
```
Comment (fritzo) [inline torch/distributions/transforms.py]: This should always return `x.pow(1 / self.exponent)`. In the non-bijective case we still want to return a pseudoinverse, as do other transforms.

**Comment by fritzo** [comment]
Actually it might be more useful to define
```py
domain = constraints.positive
codomain = constraints.positive
bijective = True
def log_abs_det_jacobian(self, x, y):
    return self.exponent * y / x
```
@alicanb what do you think?

**Comment by alicanb** [comment]
We can constrain it to positive domain in this PR, that's the place where all real exponents plays nicely. I guess this will also be equivalent to @vishwakftw 's original suggestion
> `X --> Y = log(X)` --> `affine(Y, loc=0, scale=power) = Z` -->`exp(Z) = X**power`

I really don't want to go for the subclassing option unless we have to, I think it would confuse the users.

**Comment by vishwakftw** [comment]
However, @alicanb now `PowerTransform` cannot be used for distributions like `Normal` and `Laplace`. Is that fine?

**Comment by alicanb** [comment]
@vishwakftw I know, obviously it's not ideal. I'm trying to come up with use cases for `powerTransform`, aside from https://github.com/probtorch/pytorch/issues/115, do you have any ideas?

**Comment by vishwakftw** [comment]
@alicanb I think `PowerTransform` will help complete the suite of transformations that are present as specified in https://github.com/probtorch/pytorch/issues/115 . Other than that I think `PowerTransform` could be useful for implementing other versions of distributions. For instance, `X ~ StudentT` with DOF = n, then `X.pow(2) ~ F(1, n)`. You will also be allowed to create the generalized Gamma distribution as well, IIRC.

**Comment by vishwakftw** [comment]
@fritzo Should the tests be changed to reflect the changes in the Tensor - Variable merge?

**Comment by fritzo** [comment]
> Should the tests be changed to reflect the changes in the Tensor - Variable merge?

Yes, I guess nobody had had time to update them yet.

**Comment by vishwakftw** [comment]
@fritzo @alicanb Are we going with this?

File: torch/distributions/transforms.py [hunk 2]
```diff
@@ -298,9 +299,40 @@ def log_abs_det_jacobian(self, x, y):
...
+    sign = +1
+
+    def __init__(self, exponent, cache_size=0):
+        super(PowerTransform, self).__init__(cache_size=cache_size)
+        self.exponent = broadcast_all(exponent)[0]
+
+    def __eq__(self, other):
+        if not isinstance(other, PowerTransform):
+            return False
+        result = self.exponent.eq(other.exponent).all()
```
Comment (fritzo) [inline torch/distributions/transforms.py]: I think you can now simply
```
return self.exponent.eq(other.exponent).all().item()
```

File: test/test_distributions.py [hunk 3]
```diff
@@ -2937,6 +2938,8 @@ def setUp(self):
             transforms = [
                 AbsTransform(cache_size=cache_size),
                 ExpTransform(cache_size=cache_size),
+                PowerTransform(exponent=2,
```
Comment (fritzo) [inline test/test_distributions.py]: Could you also add a tensor example:
```py
PowerTransform(torch.tensor(5).normal_().exp(), cache_size=cache_size)
```

**Comment by fritzo** [comment]
I'd merely mention Kumaraswamy in the PR description upstream, and submit a separate PR for that.

**Comment by vishwakftw** [comment]
Moving upstream. Thanks for the suggestions @fritzo @alicanb

**Pull request closed without merging**